feat: support periodic ping to principal for keeping connection alive

jparsai opened the pull request:
This PR is to introduce a configurable periodic ping to keep connection alive between agent and principal using [grpc keepalive feature](https://grpc.io/docs/guides/keepalive/). 

By default it is disabled and to enable it you need to set value of `keep-alive-ping-interval` in startup script or set env variable `KEEP_ALIVE_PING_INTERVAL`  at agent side and at principal side you have to set value of `wait-duration-for-agent-ping` in startup script or set value of env variable `WAIT_DURATION_FOR_AGENT_PING` .

Fixes https://github.com/argoproj-labs/argocd-agent/issues/171

chetan-rns commented:
gRPC has support for configuring keepalives. Do you think we should use it?

https://grpc.io/docs/guides/keepalive/

jparsai commented:
> gRPC has support for configuring keepalives. Do you think we should use it?
> 
> https://grpc.io/docs/guides/keepalive/

@chetan-rns addressed your comment PTAL

jannfis commented on cmd/agent/main.go:
```diff
@@ -176,6 +182,9 @@ func NewAgentRunCommand() *cobra.Command {
 	command.Flags().IntVar(&metricsPort, "metrics-port",
 		env.NumWithDefault("ARGOCD_AGENT_METRICS_PORT", cmd.ValidPort, 8181),
 		"Port the metrics server will listen on")
+	command.Flags().StringVar(&keepAlivePingInterval, "keep-alive-ping-interval",
```
The cobra framework has flags that take duration strings as arguments (`Duration`, `DurationVar`, etc). Using these might remove the need for parsing the duration string yourself.

jparsai replied:
Updated it, but we have to parse for env variable. PTAL

jannfis commented on cmd/cmdutil/utils.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/sirupsen/logrus"
+)
+
+func ParseKeepAlivePingInterval(interval string) time.Duration {
```
It might make sense to make this function more generic and part of the [internal/env](https://github.com/argoproj-labs/argocd-agent/blob/main/internal/env/env.go) package, so that it streamlines with the rest of the code base.

jannfis commented on cmd/principal/main.go:
```diff
@@ -275,6 +280,9 @@ func NewPrincipalRunCommand() *cobra.Command {
 		env.BoolWithDefault("ARGOCD_PRINCIPAL_ENABLE_RESOURCE_PROXY", true),
 		"Whether to enable the resource proxy")
+	command.Flags().DurationVar(&waitDurationForAgentPing, "wait-duration-for-agent-ping",
+		cmdutil.ParseKeepAlivePingInterval(env.StringWithDefault("WAIT_DURATION_FOR_AGENT_PING", nil, "")),
```
I think `WAIT_DURATION_FOR_AGENT_PING` is not a good name for the environment variable. PTAL at how the other environment variables that we use as arguments are named, and make this one align.

jannfis commented on cmd/agent/main.go:
```diff
@@ -176,6 +183,9 @@ func NewAgentRunCommand() *cobra.Command {
 		env.NumWithDefault("ARGOCD_AGENT_METRICS_PORT", cmd.ValidPort, 8181),
 		"Port the metrics server will listen on")
+	command.Flags().DurationVar(&keepAlivePingInterval, "keep-alive-ping-interval",
+		cmdutil.ParseKeepAlivePingInterval(env.StringWithDefault("KEEP_ALIVE_PING_INTERVAL", nil, "")),
```
I think `KEEP_ALIVE_PING_INTERVAL` is not a good name for the environment variable. PTAL at how the other environment variables that we use as arguments are named, and make this one align.

jannfis reviewed (commented):
I have some comments, and a question: What exactly is the `wait-duration-for-agent-ping` doing, why does it need to be set?

jparsai commented:
> I have some comments, and a question: What exactly is the `wait-duration-for-agent-ping` doing, why does it need to be set?

It is the value user wants to set for `keepalive.EnforcementPolicy.MinTime`, it is a minimum amount of time a client should wait before sending next keepalive ping. if client sends more than one pings in this window then connection will be closed by server. By default it is set to 5 minutes. It could be ignored/removed, but in case a user configures agent's `keep-alive-ping-interval` flag to less than 5 minutes (which I guess would be rare) then connection will keep on closing, so to be on safe side I thought of asking users to configure this flag as well.

Addressed your review comments PTAL.

jannfis commented on internal/env/env.go:
```diff
@@ -139,3 +142,16 @@ func StringSlice(key string, validator func(string) error) ([]string, error) {
 	return ret, nil
 }
+
+func ParseDuration(dr string) time.Duration {
```
This should directly parse a duration from the environment, much like the other functions in this package do it for other data types. 

We need two functions: One that returns a default value, and one that doesn't. So the function prototypes should be:

```go
Duration(key string, validator func(time.Duration) error) (time.Duration, error)
```
and
```go
DurationWithDefault(key string, validator func(time.Duration) error, def time.Duration) time.Duration
```

jparsai replied:
oh, my bad, I misunderstood previous comment and didn't check implementation of other functions for reference.
Addressed all your comments, PTAL

jannfis commented on cmd/principal/main.go:
```diff
@@ -275,6 +280,9 @@ func NewPrincipalRunCommand() *cobra.Command {
 		env.BoolWithDefault("ARGOCD_PRINCIPAL_ENABLE_RESOURCE_PROXY", true),
 		"Whether to enable the resource proxy")
+	command.Flags().DurationVar(&waitDurationForAgentPing, "wait-duration-for-agent-ping",
+		env.ParseDuration(env.StringWithDefault("ARGOCD_PRINCIPAL_WAIT_DURATION_FOR_AGENT_PING", nil, "")),
```
Please take a look at my comment for the new function in the `env` package to see what should be changed here.

jannfis commented on cmd/principal/main.go:
```diff
@@ -275,6 +280,9 @@ func NewPrincipalRunCommand() *cobra.Command {
 	command.Flags().BoolVar(&enableResourceProxy, "enable-resource-proxy",
 		env.BoolWithDefault("ARGOCD_PRINCIPAL_ENABLE_RESOURCE_PROXY", true),
 		"Whether to enable the resource proxy")
+	command.Flags().DurationVar(&waitDurationForAgentPing, "wait-duration-for-agent-ping",
```
With the explanation that you gave me for the purpose of this switch, I think the phrase "wait duration" might be a bit misleading here. To me, this phrase implies that the principal waits some time for the agent ping, when in reality it enforces agent pings to not being sent at a shorter interval than the value given. Good command line option names should be more or less self-explaining.

So for this, `enforce-min-keepalive` or `keepalive-min-interval` would be two options that come to my mind. I'm open for suggestions of other names.

jannfis commented on cmd/principal/main.go:
```diff
@@ -275,6 +280,9 @@ func NewPrincipalRunCommand() *cobra.Command {
 		"Whether to enable the resource proxy")
+	command.Flags().DurationVar(&waitDurationForAgentPing, "wait-duration-for-agent-ping",
+		env.ParseDuration(env.StringWithDefault("ARGOCD_PRINCIPAL_WAIT_DURATION_FOR_AGENT_PING", nil, "")),
+		"Minimum time duration agent should wait before next ping to principal") // It should be less than "keep-alive-ping-interval" of agent
```
Since this option will have dire consequences for agents (e.g. their connections will be terminated if they don't abide by this enforcement rule), we should call that out in the option's description. For example, `Drop agent connections that send keepalives more often than the specified interval`

jannfis commented on principal/options.go:
```diff
@@ -377,3 +377,10 @@ func WithResourceProxyTLS(tlsConfig *tls.Config) ServerOption {
 	}
 }
+
+func WithWaitDurationForAgentPing(interval time.Duration) ServerOption {
```
Similar comment as for the command line. This should be renamed to be more in line with what it does.

jannfis reviewed (commented):
We're coming close :)

jannfis commented on internal/env/env.go:
```diff
@@ -139,3 +140,32 @@ func StringSlice(key string, validator func(string) error) ([]string, error) {
 	return ret, nil
 }
+
+func Duration(key string) (time.Duration, error) {
```
The `validator` argument is missing from this func, as well as its execution.

jparsai replied:
ah, in hurry I missed it :(

jannfis reviewed (commented):
Cool. I have one additional thing, and also unit tests for the new code in the env package is missing (it's perfectly testable).
